### ISSUE TITLE
Fast reader in io.ascii does not work with multiprocessing 'spawn' method

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#8851]
The [multiprocessing module in Python](https://docs.python.org/3/library/multiprocessing.html) supports several modes - including ``fork`` and ``spawn``. However, currently things don't work properly with the ``spawn`` method:

```python
import multiprocessing as mp
from astropy.io.ascii import read

if __name__ == "__main__":

    mp.set_start_method('spawn')
    print(read('a,b\n1,2\n3,4\n5,\n6,7', fast_reader={'parallel': True}))
```

gives:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
  File "astropy/io/ascii/cparser.pyx", line 836, in astropy.io.ascii.cparser._copy_cparser
  File "astropy/io/ascii/cparser.pyx", line 192, in astropy.io.ascii.cparser.CParser.__cinit__
TypeError: __cinit__() got an unexpected keyword argument 'expchar'
...
```

This will cause issues for Python 3.8 since the default method on MacOS X is changing from ``fork`` to ``spawn`` (see https://github.com/python/cpython/commit/17a5588740b3d126d546ad1a13bdac4e028e6d50). The easy fix is of course to hard-code the method to ``fork`` in the fast reader, but this doesn't really fix the underlying issue. It would be good to fix this properly since there is probably a good reason the Python dev team switched to spawn (their changelog entry says '  On macOS, *spawn* start method is now the default: *fork* start method is no
   longer reliable on macOS, see https://bugs.python.org/issue33725.')

In any case the above issue can be reproduced with existing Python versions, so there's definitely a bug here to fix.

**Comment by astrofrog** [comment]
https://github.com/astropy/astropy/pull/8852 shows a workaround for now - note that once that PR is merged, the workaround will need to be removed manually to reproduce the above issue.

**Comment by astrofrog** [comment]
I started working on a proper fix in the following branch: https://github.com/astropy/astropy/compare/master...astrofrog:fix-io-ascii-38 - however I got stuck as I get errors such as:

```
Python(27389,0x118d455c0) malloc: *** error for object 0x10100000100: pointer being freed was not allocated
Python(27389,0x118d455c0) malloc: *** set a breakpoint in malloc_error_break to debug
Python(27388,0x10d3a15c0) malloc: *** error for object 0x7fb937df9c30: pointer being freed was not allocated
Python(27388,0x10d3a15c0) malloc: *** set a breakpoint in malloc_error_break to debug
Python(27390,0x10e1515c0) malloc: *** error for object 0x7f8c3b5f9560: pointer being freed was not allocated
Python(27390,0x10e1515c0) malloc: *** set a breakpoint in malloc_error_break to debug
```

It would be great if someone more familiar with C array allocation could take a look!

**Comment by MSeifert04** [comment]
Hm, did you have time to find the root cause? I wonder if this is a problem with our code or CPython/Cython? Just curious :)

**Comment by astrofrog** [comment]
@MSeifert04 - I haven't got any further than the fixes in #8853, and won't have time for the near future. But at least it's easily reproducible with all Python versions, not just 3.8.

**Comment by astrofrog** [comment]
@MSeifert04 - just for info, the malloc warnings above seem to come from ``delete_tokenizer``

**Comment by astrofrog** [comment]
Interestingly on Windows 'spawn' is already the default, and we apparently skip the test saying parallel isn't supported on Windows, maybe because of the same issues!

https://github.com/astropy/astropy/blob/master/astropy/io/ascii/tests/test_c_reader.py#L75

**Comment by astrofrog** [comment]
Just for info, if I apply the changes in #8853, I see memory allocation issues on Linux too:

```
free(): double free detected in tcache 2
free(): double free detected in tcache 2
free(): double free detected in tcache 2
free(): double free detected in tcache 2
 a   b 
--- ---
  1   2
  3   4
  5  --
  6   7
```

**Comment by astrofrog** [comment]
I have a fix!

**Comment by astrofrog** [comment]
Complete fix in https://github.com/astropy/astropy/pull/8853

**Comment by saimn** [comment]
Fixed by #8853, thanks @astrofrog :tada:
